genesis/sensors/data_recorder.py

```diff
@@ -4,191 +4,133 @@
-            self._processor_threads[i].join()
 
-    def step(self):
+        for recording_idx, recording in enumerate(self._recordings):
+            if recording.run_in_thread and self._processor_threads[recording_idx]:
+                self._processor_threads[recording_idx].join()
+            recording.handler.cleanup()
+
+    def step(self, global_step: int, global_time: float):
         """
-        Increment the step count and process sensor data if recording is active.
-        Each sensor is read every n steps based on its associated RecordingOptions.
+        Increment the step count and process data from each recording configuration.
+        In threaded mode, data is put in queues. In non-threaded mode, data is processed synchronously.
         """
-        self._step += 1
-        if not self._is_recording or self._is_paused:
+        if not self._is_recording:
             return
-        for opt in self._rec_options:
-            if self._step % opt._steps_per_sample == 0:
-                self.read_sensor(opt._sensor_idx)
 
-    def read_sensor(self, rec_idx=0):
-        """
-        Read and process data from the sensor.
-        Usually called by step(), but can also be called manually to process data immediately.
-        """
-        options = self._rec_options[rec_idx]
-        data = self._sensors[options._sensor_idx].read()
-        data_queue = self._data_queues[rec_idx]
-        try:
-            data_queue.put(data, block=True, timeout=options.buffer_full_wait_time)
-        except queue.Full:
-            gs.logger.warning("Data queue is full, dropping oldest data sample.")
-            data_queue.get_nowait()
-            data_queue.put_nowait(data)
-
-    def make_process_callback(self, rec_idx):
+        for recording_idx, recording in enumerate(self._recordings):
+            if global_step % recording._steps_per_sample == 0:
+                data = recording.data_func()
+
+                if recording.run_in_thread:
+                    data_and_time = (data, global_time)
+                    data_queue = self._data_queues[recording_idx]
+                    if data_queue is not None:
+                        try:
+                            data_queue.put(data_and_time, block=False)
+                        except queue.Full:
+                            try:
+                                data_queue.put(data_and_time, block=True, timeout=recording.buffer_full_wait_time)
+                            except queue.Full:
+                                gs.logger.debug("Data queue is full, dropping oldest data sample.")
+                                try:
+                                    data_queue.get_nowait()
+                                    data_queue.put_nowait(data_and_time)
+                                except queue.Empty:
+                                    # Queue became empty between operations, just put the data
+                                    data_queue.put_nowait(data_and_time)
```

Comment:
There is no way to avoid 3 nested level? it would be great to flatten this a bit.